exonum-java-binding/core/src/test/java/com/exonum/binding/core/storage/indices/BaseProofMapIndexProxyIntegrationTest.java

```diff
@@ -0,0 +1,806 @@
+  }
+
+  /**
+   * Create a proof key of length 32 with the specified suffix.
+   *
+   * @param suffix a key suffix. Must be shorter than or equal to 32 bytes in UTF-8.
+   * @return a key, starting with zeroes and followed by the specified suffix encoded in UTF-8
+   */
+  private static HashCode createProofKey(String suffix) {
+    byte[] suffixBytes = bytes(suffix);
+    return createProofKey(suffixBytes);
+  }
+
+  private static HashCode createProofKey(byte... suffixBytes) {
+    byte[] proofKey = createRawProofKey(suffixBytes);
+    return HashCode.fromBytes(proofKey);
+  }
+
+  private static byte[] createRawProofKey(byte... suffixBytes) {
+    checkArgument(suffixBytes.length <= PROOF_MAP_KEY_SIZE);
+    byte[] proofKey = new byte[PROOF_MAP_KEY_SIZE];
+    System.arraycopy(suffixBytes, 0, proofKey, PROOF_MAP_KEY_SIZE - suffixBytes.length,
+        suffixBytes.length);
+    return checkProofKey(proofKey);
+  }
+
+  void runTestWithView(Function<Cleaner, View> viewFactory,
+      Consumer<ProofMapIndexProxy<HashCode, String>> mapTest) {
+    runTestWithView(viewFactory, (ignoredView, map) -> mapTest.accept(map));
+  }
+
+  private void runTestWithView(
+      Function<Cleaner, View> viewFactory,
+      BiConsumer<View, ProofMapIndexProxy<HashCode, String>> mapTest) {
+    try (Cleaner cleaner = new Cleaner()) {
+      View view = viewFactory.apply(cleaner);
+      ProofMapIndexProxy<HashCode, String> map = this.create(MAP_NAME, view);
+
+      mapTest.accept(view, map);
+    } catch (CloseFailuresException e) {
+      throw new AssertionError("Unexpected exception", e);
+    }
+  }
+
+  @Override
+  StorageIndex createOfOtherType(String name, View view) {
+    return ListIndexProxy.newInstance(name, view, StandardSerializers.string());
+  }
+
+  @Override
+  Object getAnyElement(ProofMapIndexProxy<HashCode, String> index) {
+    return index.get(key1);
+  }
+
+  @Override
+  void update(ProofMapIndexProxy<HashCode, String> index) {
+    index.put(key1, V1);
+  }
+
+  /**
+   * Creates `numOfEntries` map entries, sorted by key:
+   * [(00…0PK1, V1), (00…0PK2, V2), … (00…0PKi, Vi)].
+   */
+  List<MapEntry<HashCode, String>> createSortedMapEntries() {
+    // Use PROOF_KEYS which are already sorted.
+    return createMapEntries(getProofKeys().stream());
+  }
+
+  /**
+   * Creates map entries with the given keys. Uses values
+   * from {@linkplain TestStorageItems#values} in a round-robin fashion.
+   */
+  private List<MapEntry<HashCode, String>> createMapEntries(Stream<HashCode> proofKeys) {
+    Stream<HashCode> keys = proofKeys.distinct();
+    Stream<String> roundRobinValues = IntStream.range(0, Integer.MAX_VALUE)
+        .mapToObj(i -> values.get(i % values.size()));
+    return Streams.zip(keys, roundRobinValues, MapEntry::valueOf)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Creates 257 entries for a ProofMap that, when added to it, will make the underlying
```

Comment:
Tests that rely on keys being equal to proof paths make no sense in the base but
non-hashing flavour of the ProofMap where this property holds. In hashing variant
there is no apparent connection between a key bits and the bits in the corresponding proof path. 

The tests that set particular key bits to achieve a certain tree structure to exercise the proof validation must be moved to non-hashing variant.